packages/dds/tree/src/simple-tree/schemaFactory.ts

```diff
@@ -365,17 +369,15 @@ export class SchemaFactory<
 				const customizable = this.constructor !== schema;
 				const proxyTarget = customizable ? this : undefined;
 
-				if (isFlexTreeNode(input)) {
-					return createNodeProxy(input, customizable, proxyTarget) as schema;
-				} else {
-					const flexSchema = getFlexSchema(this.constructor as TreeNodeSchema);
-					return createRawNodeProxy(
-						flexSchema as FlexObjectNodeSchema,
-						input,
-						customizable,
-						proxyTarget,
-					) as unknown as schema;
-				}
+				const flexSchema = getFlexSchema(this.constructor as TreeNodeSchema);
+				assert(flexSchema instanceof FlexObjectNodeSchema, "invalid flex schema");
+				const flexNode: FlexTreeNode = isFlexTreeNode(input)
+					? input
+					: createRawNode(flexSchema, copyContent(flexSchema.name, input) as object);
+
+				const proxy: TreeNode = createObjectProxy(flexSchema, customizable, proxyTarget);
+				setFlexNode(proxy, flexNode);
```

Comment:
I think the responsibility of associating the flex node with the proxy should be done by the inner functions (`createObjectProxy`, `createMapProxy`, `createArrayNodeProxy`) so that a caller can never forget to do it. It used to be enforced as such because these functions were private and you had to go through the outer catch-all function (`createRawNodeProxy`) which always called `setFlexNode`. But now, the three inner function are exposed. Which is good, but they should probably do the flex node setting instead of the caller.

I think the same probably applies to `copyContent`. In fact, since you are getting rid of runtime dispatch here and hardcoding to each type of proxy, can we just get rid of `copyContent` and inline the correct kind of copy (and the forwarding of the type name symbol) inside each of `createObjectProxy`, `createMapProxy`, and `createArrayNodeProxy`?

Reply:
The inner creation functions don't currently have access to the flex-tree nodes, and passing in the node gets a bit sketchy since if they close over it and use the passed in one somewhere then thats a bug as it will break hydration. GIven these functions are called in exactly one place now, I don't think its really an issue for setting the flex tree node to be in the constructor instead of this helper. Longer term I think we should refactor this code so that we have one file per node kind in simple-tree and remove this arbitrary factory vs proxies split since the fact that some (currently all) node kinds use proxies really should be an implementation detail of that specific node.

As for copy content, I don't really understand how the typeNameSymbol property it adds to the copy is used, and doing that requires treating it as a arbitrary object anyway. I do agree that should be cleaned up, but I think its out of scope for this PR as it touches hydration I stuff I don't understand and rather not mix up into this simple refactor.

Reply:
Okay. I'll think about how this can be cleaned up further later after the changes to hydration settle.